### PR TITLE
feat: add opt-in diffusion tuning cache prefetch

### PR DESCRIPTION
**Pull request by changwangss**
## Motivation                                                                                                                                                                                             
                                                                                                                                                                                                            
  Repeated CPU-to-GPU cache transfers can become a bottleneck during Wan2.2 tuning with `low_gpu_mem_usage=True`.

  For Wan2.2 T2V-A14B at 480×832 resolution and 81 frames:
  - Video tokens: 21 × 30 × 52 = 32,760.
  - One BF16 hidden-state tensor: 32,760 × 5,120 × 2 bytes ≈ 0.312 GiB.
  - A cached block input and reference output together occupy approximately 0.625 GiB per sample, assuming both are BF16.
  - For 128 cached samples, these tensors alone require approximately 80 GiB, excluding conditioning tensors, model weights, and tuning activations.

  Keeping these caches on CPU reduces GPU memory usage, but repeatedly transferring selected batches can leave the GPU waiting for data.


## Description
  Add `diffusion_tuning_cache_size` to enable bounded prefetching with two reusable buffers and optional GPU storage for best-parameter snapshots. Set it to `"auto"` to determine the budget after warmup,
  or specify a budget in GiB. Insufficient memory falls back to the existing path.

  Disabled by default and limited to single-GPU diffusion SignRound tuning.

  Usage example (single-GPU Wan2.2 T2V-A14B):

  ```bash
  CUDA_VISIBLE_DEVICES=0 auto-round \
      --model Wan-AI/Wan2.2-T2V-A14B-Diffusers \
      --scheme MXFP4 \
      --model_dtype bf16 \
      --format fake \
      --batch_size 1 \
      --dataset coco2014 \
      --calib_num_inference_steps 8 \
      --low_gpu_mem_usage \
      --diffusion_tuning_cache_size auto \
      --output_dir ./wan2.2-a14b-mxfp4
  ```

  `auto` selects the extra GPU cache budget after the first tuning iteration. Alternatively, use a numeric value such as `8` for an 8 GiB cache budget. This budget applies to the additional cache buffers,
  not total GPU memory usage.

  The feature requires `low_gpu_mem_usage`. Omit `diffusion_tuning_cache_size` or set it to `0` to retain the existing behavior.

## performance improve with wan2.2 T2V
device: b300

main branch, quantize transformer about need 5 hours 44 mins.
```
CUDA_VISIBLE_DEVICES=0  auto_round   --model /models/Wan2.2-T2V-A14B-Diffusers   --model_dtype bf16   --scheme MXFP4   --iters 50   --nsamples 32   --batch_size 1    --calib_num_inference_steps 8   --format llm_compressor   --device 0   --disable_low_cpu_mem_usage   --output_dir /models/changwa1/wan22-mxfp4-n32-i50   --low_gpu_mem_usage 
```
```
2026-09-10 02:07:13 INFO orchestrator.py L619: caching done
Quantizing blocks.0:   0%|                                                                               | 0/40 [00:06<?, ?it/s]
quantized 10/10 layers in the block, loss iter 0: 3.508e-03 -> iter 48: 6.870e-04
2026-09-10 02:17:52 INFO device.py L1450: 'peak_ram': 619.49GB, 'peak_vram': 28.9GB
Quantizing blocks.1:   2%|█▋                                                                  | 1/40 [10:47<7:01:00, 647.70s/it]
quantized 10/10 layers in the block, loss iter 0: 4.217e-03 -> iter 35: 2.298e-03
2026-09-10 02:27:47 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
Quantizing blocks.2:   5%|███▍                                                                | 2/40 [20:34<6:27:26, 611.75s/it]
quantized 10/10 layers in the block, loss iter 0: 1.036e-02 -> iter 49: 3.589e-03
2026-09-10 02:37:16 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
Quantizing blocks.3:   8%|█████                                                               | 3/40 [30:02<6:05:06, 592.07s/it]
quantized 10/10 layers in the block, loss iter 0: 6.333e-03 -> iter 43: 4.098e-03
...
Quantizing blocks.37:  92%|█████████████████████████████████████████████████████████████     | 37/40 [5:16:13<25:55, 518.51s/it]
quantized 10/10 layers in the block, loss iter 0: 1.815e-01 -> iter 42: 7.170e-02
2026-09-10 07:32:53 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
Quantizing blocks.38:  95%|██████████████████████████████████████████████████████████████▋   | 38/40 [5:25:40<17:45, 532.96s/it]
quantized 10/10 layers in the block, loss iter 0: 3.492e-01 -> iter 46: 4.190e-02
2026-09-10 07:42:23 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
Quantizing blocks.39:  98%|████████████████████████████████████████████████████████████████▎ | 39/40 [5:35:10<09:04, 544.10s/it]
quantized 10/10 layers in the block, loss iter 0: 6.446e-02 -> iter 39: 4.421e-02
2026-09-10 07:51:25 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
Quantizing done: 100%|███████████████████████████████████████████████████████████████████████| 40/40 [5:44:36<00:00, 516.90s/it]
2026-09-10 07:51:49 INFO device.py L1450: 'peak_ram': 632.47GB, 'peak_vram': 28.9GB
2026-09-10 07:51:49 INFO orchestrator.py L808: quantization tuning time 20676.152049064636
2026-09-10 07:51:49 INFO orchestrator.py L827: Summary: quantized 401/407 in the model, unquantized layers: condition_embedder.text_embedder.linear_1, condition_embedder.text_embedder.linear_2, condition_embedder.time_embedder.linear_1, condition_embedder.time_embedder.linear_2, condition_embedder.time_proj, proj_out
```
with this pr, reduce to about 2 hours 45 min.
```
CUDA_VISIBLE_DEVICES=1  auto_round   --model /models/Wan2.2-T2V-A14B-Diffusers   --model_dtype bf16   --scheme MXFP4   --iters 50   --nsamples 32   --batch_size 1    --calib_num_inference_steps 8   --format llm_compressor   --device 0   --disable_low_cpu_mem_usage   --output_dir /models/changwa1/wan22-mxfp4-n32-i50-prefetch   --low_gpu_mem_usage --diffusion_tuning_cache_size auto
```
```
2026-09-10 10:51:12 INFO orchestrator.py L619: caching done
Quantizing blocks.0:   0%|                                                                               | 0/40 [00:05<?, ?it/s]2026-09-10 10:53:41 INFO tuning_cache.py L92: Diffusion tuning auto cache budget after warmup: 120.20 GiB.
2026-09-10 10:53:42 INFO tuning_cache.py L122: Diffusion tuning prefetch enabled with 2.68 GiB of GPU buffers; best parameters on GPU.
quantized 10/10 layers in the block, loss iter 0: 3.771e-03 -> iter 48: 3.993e-04
2026-09-10 10:55:29 INFO device.py L1560: 'peak_ram': 256.99GB, 'peak_vram': 28.11GB
Quantizing blocks.1:   2%|█▋                                                                  | 1/40 [04:23<2:51:10, 263.34s/it]
quantized 10/10 layers in the block, loss iter 0: 1.610e-02 -> iter 30: 1.638e-03
2026-09-10 10:59:34 INFO device.py L1560: 'peak_ram': 268.48GB, 'peak_vram': 28.15GB
Quantizing blocks.2:   5%|███▍                                                                | 2/40 [08:22<2:37:42, 249.01s/it]
quantized 10/10 layers in the block, loss iter 0: 6.369e-03 -> iter 42: 2.355e-03
2026-09-10 11:04:08 INFO device.py L1560: 'peak_ram': 268.48GB, 'peak_vram': 28.15GB
Quantizing blocks.3:   8%|█████                                                               | 3/40 [12:56<2:40:35, 260.43s/it]
quantized 10/10 layers in the block, loss iter 0: 5.187e-03 -> iter 48: 3.225e-03
...
Quantizing blocks.38:  95%|██████████████████████████████████████████████████████████████▋   | 38/40 [2:38:22<07:47, 233.64s/it]
quantized 10/10 layers in the block, loss iter 0: 1.369e-01 -> iter 45: 3.600e-02
2026-09-10 13:33:10 INFO device.py L1560: 'peak_ram': 268.48GB, 'peak_vram': 28.15GB
Quantizing blocks.39:  98%|████████████████████████████████████████████████████████████████▎ | 39/40 [2:41:58<03:48, 228.30s/it]
quantized 10/10 layers in the block, loss iter 0: 1.710e-01 -> iter 39: 3.392e-02
2026-09-10 13:36:50 INFO device.py L1560: 'peak_ram': 268.48GB, 'peak_vram': 28.15GB
Quantizing done: 100%|███████████████████████████████████████████████████████████████████████| 40/40 [2:45:51<00:00, 248.78s/it]
2026-09-10 13:37:03 INFO device.py L1560: 'peak_ram': 268.48GB, 'peak_vram': 28.15GB
2026-09-10 13:37:03 INFO orchestrator.py L808: quantization tuning time 9951.35483789444
2026-09-10 13:37:03 INFO orchestrator.py L827: Summary: quantized 401/407 in the model, unquantized layers: condition_embedder.text_embedder.linear_1, condition_embedder.text_embedder.linear_2, condition_embedder.time_embedder.linear_1, condition_embedder.time_embedder.linear_2, condition_embedder.time_proj, proj_out

```


## Type of Change

<!-- Bug fix / New feature / Documentation / Performance / Refactor / Other: __________ -->

New feature

## Related Issues

<!-- Link to related issues using #issue_number -->

Fixes or relates to #2279 

## Checklist Before Submitting

- [ ] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [ ] New or updated tests are included where applicable.
- [ ] The CUDA CI has passed. You can trigger it by commenting `/azp run Unit-Test-CUDA-AutoRound`.

<!-- Optional: Tag reviewers or add extra notes below -->

### CODE REVIEW

**Comment by n1ck-guo** [comment]
Could you please explain the specific scope of this PR's impact?

File: auto_round/algorithms/quantization/sign_round/quantizer.py [hunk 1]
```diff
@@ -473,61 +473,107 @@ def quantize_block(
...
-
-                self._scale_loss_and_backward(scaler, loss)
-
-                if mid_iter_mem_check:
-                    # clear memory to avoid OOM due to memory fragmentation
-                    clear_memory_if_reached_threshold(threshold=0.8, device_list=device_manager.device_list)
+        tuning_cache = None
+        # Only opt-in diffusion tuning can enter the CUDA staging path.
+        cache_budget = getattr(self.model_context, "diffusion_tuning_cache_size", 0)
+        use_tuning_cache = (
```
Comment (n1ck-guo) [inline auto_round/algorithms/quantization/sign_round/quantizer.py:479]: I suggest setting `use_tuning_cache` as a parameter in `compress_context` and making the judgment within `compress_context`, which would make it easier to reuse

File: auto_round/algorithms/quantization/sign_round/quantizer.py [hunk 2]
```diff
@@ -473,61 +473,107 @@ def quantize_block(
...
-                    clear_memory_if_reached_threshold(threshold=0.5, device_list=device_manager.device_list)
-
-                self._scale_loss_and_backward(scaler, loss)
-
-                if mid_iter_mem_check:
-                    # clear memory to avoid OOM due to memory fragmentation
-                    clear_memory_if_reached_threshold(threshold=0.8, device_list=device_manager.device_list)
+        tuning_cache = None
+        # Only opt-in diffusion tuning can enter the CUDA staging path.
+        cache_budget = getattr(self.model_context, "diffusion_tuning_cache_size", 0)
```
Comment (n1ck-guo) [inline auto_round/algorithms/quantization/sign_round/quantizer.py:478]: maybe better set diffusion_tuning_cache_size in compress_contesxt?

File: test/unit/common/compressors/test_diffusion_tuning_cache.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import random
```
Comment (lvliang-intel) [inline test/unit/common/compressors/test_diffusion_tuning_cache.py:15]: add copy right here

**Review by lvliang-intel — APPROVED** [review]
LGTM

**Review by n1ck-guo — APPROVED** [review]
LGTM

**Comment by changwangss** [comment]
> Could you please explain the specific scope of this PR's impact?

  This change is limited to diffusion SignRound tuning with low_gpu_mem_usage=True, single-CUDA execution, and batch_size=1. It is opt-in through diffusion_tuning_cache_size and remains disabled by default, so non-diffusion, multi-GPU, other algorithms, and existing default flows keep the original behavior.

File: auto_round/algorithms/quantization/sign_round/quantizer.py (hunk 2 above)
Comment (changwangss) [reply auto_round/algorithms/quantization/sign_round/quantizer.py:478]: Thanks for the suggestion. I kept the cache decision in the SignRound tuning path because the current implementation is specific to diffusion calibration, single-CUDA execution, low_gpu_mem_usage=True,
  and batch_size=1. Keeping the guard local limits the impact on generic compress_context behavior. We can revisit moving it into compress_context when another tuning path needs the same cache mechanism.

File: test/unit/common/compressors/test_diffusion_tuning_cache.py (hunk 3 above)
Comment (changwangss) [reply test/unit/common/compressors/test_diffusion_tuning_cache.py:15]: improved.

File: auto_round/algorithms/quantization/sign_round/quantizer.py [hunk 4]
```diff
@@ -473,61 +473,107 @@ def quantize_block(
...
+            and len(device_manager.device_list) == 1
+            and (loss_device is None or torch.device(loss_device) == torch.device(device))
+        )
 
-            if i == 0:
-                init_loss = total_loss
-            current_lr = optimizer.param_groups[0]["lr"]
-            logger.debug("iter %d loss: %.3e lr: %s", i, total_loss, current_lr)
+        try:
+            for i in range(self.iters):
```
Comment (wenhuach21) [inline auto_round/algorithms/quantization/sign_round/quantizer.py:489]: May I know why these code should be added to here, all data code should move to block forward or diffusion. Please try your best to refine this in the following pr

File: auto_round/algorithms/quantization/sign_round/quantizer.py [hunk 5]
```diff
@@ -473,61 +473,107 @@ def quantize_block(
...
-            sync_gradients()
-            self._step(scaler, optimizer, lr_schedule)
+                    if 0 < self.dynamic_max_gap <= i - last_best_iter:
+                        break
+                sync_gradients()
+                self._step(scaler, optimizer, lr_schedule)
+
+        finally:
+            if tuning_cache is not None:
+                tuning_cache.close()
```
Comment (wenhuach21) [inline auto_round/algorithms/quantization/sign_round/quantizer.py:576]: There is no lock in this cache, so I don't see the purpose of using `try/finally` here. Since `try/finally` adds some extra indentation and makes the code slightly harder to read, I would suggest removing it unless there is a specific cleanup or correctness requirement that I'm missing.
